### PR TITLE
resources/images: Create padding image filter

### PR DESCRIPTION
**Pull request by jmooring**
This addresses the primary use case described in https://github.com/gohugoio/hugo/issues/11599.

### CODE REVIEW

**Comment by jmooring** [comment]
Can also combine with the `.Colors` method...

```text
{{ with resources.Get "images/c.jpg" }}
  {{ with .Filter (images.Padding "20" (index .Colors 2)) }}
    <img src="{{ .RelPermalink }}" width="{{ .Width }}" height="{{ .Height }}" alt="">
  {{ end }}
{{ end }}
```

File: resources/images/filters.go [hunk 1]
```diff
@@ -110,6 +113,65 @@ func (*Filters) Text(text string, options ...any) gift.Filter {
 	}
 }
 
+// Padding creates a filter that resizes the image canvas without resizing the
+// image. The last argument is the canvas color, expressed as an RGB or RGBA
+// hexadecimal color. The default value is `ffffffff` (opaque white). The
+// preceding arguments are the padding values, in pixels, using the CSS
+// shorthand property syntax. Negative padding values will crop the image. The
+// signature is images.Padding V1 [V2] [V3] [V4] [COLOR].
+func (*Filters) Padding(args ...any) gift.Filter {
```
Comment (bep) [inline resources/images/filters.go:122]: I like the overall idea of the Padding func. But I think the variadic function setup becomes too complicated with 5 different arguments.

I suggest

* using a dict with top, right, bottom, left, color
* create a struct with these and use mapstructure.WeakDecode
* But make sure to pass the original dict (the one provided from the user) to newFilterOpts (this allows us to add fields to the options struct without breaking the hash)

File: resources/images/padding.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
+// Copyright 2021 The Hugo Authors. All rights reserved.
+//
```
Comment (bep) [inline resources/images/padding.go:2]: 2023.

**Review by bep — CHANGES_REQUESTED** [review]

File: resources/images/filters.go (hunk 1 above)
Comment (jmooring) [reply resources/images/filters.go:122]: This is a polite request to reconsider. In my view the most common cases are:

```go-html-template
{{ $filter := images.Padding 20 }}
{{ $filter := images.Padding 20 40 }}
```

Also, the CSS padding shorthand should be familiar to most site and theme authors, and I can certainly add links to mozilla docs describing the notation.

I'm fine with the options map; it just seems like a little more work compared to some of the other filters, excluding text and overlay.

Comment (bep) [reply resources/images/filters.go:122]: Can you guarantee that there will be no more options coming in the future here?

Comment (bep) [reply resources/images/filters.go:122]: What if you could have both ... So we would check for the common cases (1 or 2 integers) as the special case.

Comment (bep) [reply resources/images/filters.go:122]: Yea, ok, we could introduce the dict if it gets more complicated.

File: resources/images/filters.go [hunk 3]
```diff
@@ -110,6 +113,65 @@ func (*Filters) Text(text string, options ...any) gift.Filter {
...
+	if len(args) == 0 {
+		panic("the padding filter requires one or more arguments")
+	}
+
+	var top, right, bottom, left int
+	var ccolor color.Color = color.White // canvas color
+	var err error
+
+	vc := args[len(args)-1]
+	if reflect.TypeOf(vc).Kind() == reflect.String {
```
Comment (bep) [inline resources/images/filters.go:132]: I don't see a need to use reflect here, do a "Go type cast".

File: resources/images/filters.go [hunk 4]
```diff
@@ -110,6 +113,65 @@ func (*Filters) Text(text string, options ...any) gift.Filter {
...
+	case 2:
+		top, right, bottom, left = vals[0], vals[1], vals[0], vals[1]
+	case 3:
+		top, right, bottom, left = vals[0], vals[1], vals[2], vals[1]
+	case 4:
+		top, right, bottom, left = vals[0], vals[1], vals[2], vals[3]
+	}
+
+	return filter{
+		Options: newFilterOpts(top, right, bottom, left, ccolor),
```
Comment (bep) [inline resources/images/filters.go:164]: Use vc instead of ccolor here. Probably not important, but will at least guarantee the most stable hash.

**Review by bep — CHANGES_REQUESTED** [review]

File: resources/images/filters.go (hunk 4 above)
Comment (bep) [inline resources/images/filters.go:164]: Just pass `args..` in here, which would make the most stable hash.

**Pull request merged**